File: valid-parentheses/valid-parentheses.cpp

class Solution {
public:
    bool isValid(string s) {
        
        stack<string> st;
        unordered_map<string, string> map;
        map[")"] = "(";
        map["]"] = "[";
        map["}"] = "{";
        
        for(int i=0; i < s.size(); i++){
            if(map.find(string(1,s[i])) != map.end() ){
                string topelem = !st.empty() ? st.top() : "#";
                if(!st.empty())
                    st.pop();

                if( map[string(1, s[i])] != string(1, topelem[0]) )
                    return false;
            }
            else
               st.push(string(1, s[i]));
        }
        return st.empty();
    }
};

File: valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        open = ['(','[','{']
        close = [')',']','}']
        for ch in s:
            if ch in open:
                stack.append(ch)
            elif ch in close:
                if stack and open.index(stack.pop()) == close.index(ch):
                    print('ffffhi')
                    pass
                else:
                    print('hfi')
                    return False
            else: 
                print('hi')
                return False
        return not stack